db/sql_implementation.go

```diff
@@ -0,0 +1,798 @@
+		case Less:
+			whereConditions[i] = sqlz.Lt(string(filterOpt.Field), value)
+		case Greater:
+			whereConditions[i] = sqlz.Gt(string(filterOpt.Field), value)
+		case LessOrEqual:
+			whereConditions[i] = sqlz.Lte(string(filterOpt.Field), value)
+		case GreaterOrEqual:
+			whereConditions[i] = sqlz.Gte(string(filterOpt.Field), value)
+		case Contains:
+			// Note(albrow): If needed, we can optimize this so it is easier to index.
+			// LIKE queries are notoriously slow.
+			whereConditions[i] = sqlz.Like(string(filterOpt.Field), fmt.Sprintf("%%%s%%", value))
+		default:
+			return nil, fmt.Errorf("db.FindMiniHeaders: unknown FilterOpt.Kind: %s", filterOpt.Kind)
+		}
+	}
+	return whereConditions, nil
+}
+
+func (db *DB) DeleteMiniHeader(hash common.Hash) error {
+	if _, err := db.sqldb.ExecContext(db.ctx, "DELETE FROM miniHeaders WHERE hash = $1", hash); err != nil {
+		return convertErr(err)
+	}
+	return nil
+}
+
+func (db *DB) DeleteMiniHeaders(query *MiniHeaderQuery) (deleted []*types.MiniHeader, err error) {
+	defer func() {
+		err = convertErr(err)
+	}()
+	// HACK(albrow): sqlz doesn't support ORDER BY, LIMIT, and OFFSET
+	// for DELETE statements. It also doesn't support RETURNING. As a
+	// workaround, we do a SELECT and DELETE inside a transaction.
+	var miniHeadersToDelete []*sqltypes.MiniHeader
+	err = db.sqldb.TransactionalContext(db.ctx, nil, func(tx *sqlz.Tx) error {
+		stmt, err := findMiniHeadersQueryFromOpts(tx, query)
+		if err != nil {
+			return err
+		}
+		if err := stmt.GetAllContext(db.ctx, &miniHeadersToDelete); err != nil {
+			return err
+		}
+		for _, miniHeader := range miniHeadersToDelete {
+			_, err := tx.DeleteFrom("miniHeaders").Where(sqlz.Eq(string(MFHash), miniHeader.Hash)).ExecContext(db.ctx)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return sqltypes.MiniHeadersToCommonType(miniHeadersToDelete), nil
+}
+
+// GetMetadata returns the metadata (or db.ErrNotFound if no metadata has been saved).
+func (db *DB) GetMetadata() (*types.Metadata, error) {
+	var metadata sqltypes.Metadata
+	if err := db.sqldb.GetContext(db.ctx, &metadata, "SELECT * FROM metadata LIMIT 1"); err != nil {
+		return nil, convertErr(err)
+	}
+	return sqltypes.MetadataToCommonType(&metadata), nil
+}
+
+// SaveMetadata inserts the metadata into the database, overwriting any existing
+// metadata. It returns ErrMetadataAlreadyExists if the metadata has already been
+// saved in the database.
+func (db *DB) SaveMetadata(metadata *types.Metadata) (err error) {
+	defer func() {
+		err = convertErr(err)
+	}()
+	err = db.sqldb.TransactionalContext(db.ctx, nil, func(txn *sqlz.Tx) error {
+		query := db.sqldb.Select("COUNT(*)").From("metadata")
+		count, err := query.GetCount()
+		if err != nil {
+			return err
+		}
+		if count != 0 {
+			return ErrMetadataAlreadyExists
+		}
+		_, err = db.sqldb.NamedExecContext(db.ctx, insertMetadataQuery, sqltypes.MetadataFromCommonType(metadata))
```

Comment:
Shouldn't we throw an error if `count == 1 too`?

Reply:
What do you mean? If `count == 1` is true than `count != 0` is also true.

Reply:
I think I misread this.